pageserver/src/tenant/timeline.rs

```diff
@@ -869,17 +872,71 @@ impl Timeline {
 
     pub async fn evict_layer(&self, layer_file_name: &str) -> anyhow::Result<Option<bool>> {
         let Some(local_layer) = self.find_layer(layer_file_name) else { return Ok(None) };
-        if local_layer.is_remote_layer() {
-            return Ok(Some(false));
-        }
         let Some(remote_client) = &self.remote_client else { return Ok(Some(false)) };
+        let results = self
+            .evict_layer_batch(remote_client, &[local_layer])
+            .await?;
+        assert_eq!(results.len(), 1);
+        let result: Option<anyhow::Result<bool>> = results.into_iter().next().unwrap();
+        match result {
+            None => anyhow::bail!("task_mgr shutdown requested"),
+            Some(Ok(b)) => Ok(Some(b)),
+            Some(Err(e)) => Err(e),
+        }
+    }
 
-        // ensure the current layer is uploaded for sure
+    // The `remote_client` should be this timeline's `self.remote_client`.
+    // We make the caller provide it so that they are responsible for handling the case
+    // where someone wants to evict the layer but no remote storage is configured.
+    async fn evict_layer_batch(
+        &self,
+        remote_client: &Arc<RemoteTimelineClient>,
+        layers_to_evict: &[Arc<dyn PersistentLayer>],
+    ) -> anyhow::Result<Vec<Option<anyhow::Result<bool>>>> {
+        // ensure that the layers have finished uploading
+        // (don't hold the layer_removal_cs while we do it, we're not removing anything yet)
         remote_client
             .wait_completion()
             .await
             .context("wait for layer upload ops to complete")?;
 
+        // now lock out layer removal (compaction, gc, timeline deletion)
+        let layer_removal_guard = self.layer_removal_cs.lock().await;
+
+        // start the batch update
+        let mut layer_map = self.layers.write().unwrap();
+        let mut batch_updates = layer_map.batch_update();
+
+        let mut results = Vec::with_capacity(layers_to_evict.len());
+
+        for l in layers_to_evict.iter() {
+            let res = if task_mgr::is_shutdown_requested() {
```

Comment:
While this right now reads a cancellation token, I do feel a bit uneasy with it.

```rust
let expected = layers_to_evict.len();

results.extend(
  layers_to_evict
    .into_iter()
    .take_while(|_| task_mgr::is_shutdown_requested())
    .map(|l| self.evict_layer_batch_impl(&guard, l, &mut batch))
    .map(Some));

// fill in the blanks
results.extend(std::iter::repeat(None).take(expected - results.len()));
```

This would read the `false` from `is_shutdown_requested` only once.

Reply:
Hm, that re-orders the results though.
Nothing relies on it right now, but still.
Let's take this discussion to https://github.com/neondatabase/neon/pull/3557

Reply:
Taken, discussed, it did not reorder, but for loop was kept due to this being quite unreadable and semi-misusing map.